### ISSUE TITLE
ReBench broken with strict YAML 1.2 parser

### ISSUE THREAD

**Issue by IsaacOscar** [smarr/ReBench#122]
Rebench no longer seems to work at all, and produces the following error (whenever I try and run `rebench somefile.yml` (for any non-empty `somefile.yml`):
```
Traceback (most recent call last):
  File "/usr/local/bin/rebench", line 11, in <module>
    load_entry_point('ReBench==1.0rc2', 'console_scripts', 'rebench')()
  File "/usr/local/lib/python3.7/site-packages/rebench/rebench.py", line 244, in main_func
    return 0 if rebench.run() else -1
  File "/usr/local/lib/python3.7/site-packages/rebench/rebench.py", line 210, in run
    config = load_config(args.config[0])
  File "/usr/local/lib/python3.7/site-packages/rebench/configurator.py", line 130, in load_config
    validator.validate(raise_exception=True)
  File "/usr/local/lib/python3.7/site-packages/pykwalify/core.py", line 155, in validate
    self._start_validate(self.source)
  File "/usr/local/lib/python3.7/site-packages/pykwalify/core.py", line 187, in _start_validate
    r = Rule(schema=v)
  File "/usr/local/lib/python3.7/site-packages/pykwalify/rule.py", line 66, in __init__
    self.init(schema, "")
  File "/usr/local/lib/python3.7/site-packages/pykwalify/rule.py", line 443, in init
    func_mapping[k](v, rule, path)
  File "/usr/local/lib/python3.7/site-packages/pykwalify/rule.py", line 1180, in init_mapping_value
    rule.init(v, u"{0}/mapping/{1}".format(path, k))
  File "/usr/local/lib/python3.7/site-packages/pykwalify/rule.py", line 443, in init
    func_mapping[k](v, rule, path)
  File "/usr/local/lib/python3.7/site-packages/pykwalify/rule.py", line 789, in init_required_value
    path=path,
pykwalify.errors.RuleError: <RuleError: error code 4: Value: 'yes' for required keyword must be a boolean: Path: '/mapping/gauge_adapter'>
```

I'm not sure when this started happening, perhaps it is caused by a update to pykwalify?

**Comment by smarr** [comment]
Hard to diagnose without further details.

Does the error `RuleError: error code 4: Value: 'yes' for required keyword must be a boolean: Path: '/mapping/gauge_adapter'` tell you anything?

There is a required schema, and "non-empty" somefile.yml does not sound like you're necessarily fulfilling that schema.

Not sure what the pykwalify update is about. But perhaps you can just use the version that ReBench worked with?

**Comment by IsaacOscar** [comment]
It's failing with every rebench file I try, such as <https://github.com/smarr/ReBench/blob/master/rebench/tests/small.conf>


I have fixed the error by running:
```
sed -i 's/yes/True/g' /usr/local/lib/python3.7/site-packages/rebench/rebench-schema.yml
sed -i 's/no/False/g' /usr/local/lib/python3.7/site-packages/rebench/rebench-schema.yml
```

pykwalify might not be at fault, as older versions have the same problem.

**Comment by smarr** [comment]
I did rerun an open PR on travis, which passed without issues: https://travis-ci.org/smarr/ReBench/builds/572530424?utm_source=github_status&utm_medium=notification

Changing the value the schema uses from [a standard YAML value](https://yaml.org/spec/1.1/#id864510) to a Python True/False seems strange.

**Comment by IsaacOscar** [comment]
Ok, It seems that my version `ruamel.yaml` is at fault,
I had version `0.15.41` installed, downgrading to `0.10.10` fixes the problem (that was the most recent version I could get working).

However, *upgrading* to `0.16.5` causes the following errors:
```
Traceback (most recent call last):
  File "/usr/local/bin/rebench", line 11, in <module>
    load_entry_point('ReBench==1.0rc1', 'console_scripts', 'rebench')()
  File "/usr/local/lib/python3.7/site-packages/rebench/rebench.py", line 248, in main_func
    return 0 if rebench.run() else -1
  File "/usr/local/lib/python3.7/site-packages/rebench/rebench.py", line 210, in run
    config = load_config(args.config[0])
  File "/usr/local/lib/python3.7/site-packages/rebench/configurator.py", line 128, in load_config
    schema_files=[dirname(__file__) + "/rebench-schema.yml"])
  File "/usr/local/lib/python3.7/site-packages/pykwalify/core.py", line 90, in __init__
    data = yaml.safe_load(stream)
  File "/usr/local/lib/python3.7/site-packages/ruamel/yaml/main.py", line 994, in safe_load
    return load(stream, SafeLoader, version)
  File "/usr/local/lib/python3.7/site-packages/ruamel/yaml/main.py", line 949, in load
    return loader._constructor.get_single_data()
  File "/usr/local/lib/python3.7/site-packages/ruamel/yaml/constructor.py", line 113, in get_single_data
    return self.construct_document(node)
  File "/usr/local/lib/python3.7/site-packages/ruamel/yaml/constructor.py", line 123, in construct_document
    for _dummy in generator:
  File "/usr/local/lib/python3.7/site-packages/ruamel/yaml/constructor.py", line 723, in construct_yaml_map
    value = self.construct_mapping(node)
  File "/usr/local/lib/python3.7/site-packages/ruamel/yaml/constructor.py", line 439, in construct_mapping
    self.flatten_mapping(node)
  File "/usr/local/lib/python3.7/site-packages/ruamel/yaml/constructor.py", line 396, in flatten_mapping
    raise DuplicateKeyError(*args)
ruamel.yaml.constructor.DuplicateKeyError: while constructing a mapping
  in "/usr/local/lib/python3.7/site-packages/rebench/rebench-schema.yml", line 151, column 5
found duplicate key "<<"
  in "/usr/local/lib/python3.7/site-packages/rebench/rebench-schema.yml", line 152, column 5

To suppress this check see:
   http://yaml.readthedocs.io/en/latest/api.html#duplicate-keys

Duplicate keys will become an error in future releases, and are errors
by default when using the new API.
```


My 

What version of `ruamel.yaml` does your travis use?

**Comment by IsaacOscar** [comment]
Hmm, it seems completely uninstalling `ruamel.yaml` fixes the problem as well.
I'll just have to remember not to install that package ever again...

(FYI I worked out that `rumel.yaml` was at fault by looking at the source for pykwalify, <https://github.com/Grokzen/pykwalify/blob/7f408193d41faaec66a718808d26c39371469862/pykwalify/compat.py#L12>`, as you can see it uses that as it's `yaml` library if possible, otherwise it fallsback to `yaml` (which seems to be PyYAML).

**Comment by smarr** [comment]
Ah, ok, thanks for figuring this out.

I'll try to set up a test for this at some point. Though, probably not in the coming weeks :(

**Comment by smarr** [comment]
(My goal would be to have it work also with ruamel.yaml present.)

**Comment by smarr** [comment]
Fixed by #123 as far as I can tell.

**Comment by IsaacOscar** [comment]
> Fixed by #123 as far as I can tell.

Thanks! I tested it now works with rumael.

**Comment by smarr** [comment]
Great, thanks for confirming!
